python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+which stores the (string) key to the Python function.
+Whenever `call_method` encounters a `pythonFunctionWrapper` object,
+it [wraps](https://github.com/mducle/hugo/blob/master/src/check_wrapped_function.m) this in a call to the mex `call_python` function with the correct key.
+Thus it appears to the user that Matlab is calling a Python function transparently (but actually happens is that Matlab calls `call_python`).
+`call_python` then [retrieves](https://github.com/mducle/hugo/blob/master/src/call_python.cpp#L390) the Python function reference from the module-global `dict` 
+and [calls](https://github.com/mducle/hugo/blob/master/src/call_python.cpp#L413) that function with the arguments given to it.
+
+The majority of the code in `call_python` is for data conversion between Matlab and Python.
+This is because `mex` functions may only take and return *Matlab* data types (Python objects are not allowed).
+Similarly the CPython `PyObject_Call` function requires that inputs to the called functions are all of the `PyObject` type.
+Thus `call_python` converts Matlab types to their Python equivalent and vice versa as per this table:
+
+|      Matlab    |      Python         | Note |
+| -------------- | ------------------- | ---- |
+| numeric array  | `numpy.ndarray`     |      |
+| numeric array  | `list` of numbers   | p2m  |
+| `char` array   | `str`               |      |
+| `string`       | `str`               |      |
+| `string` array | `list` of `str`     |      |
+| `struct`       | `dict`              |      |
+| `struct` array | `list` of `dict`    |      |
+| `cell` array   | `list`              |      |
+| `cell` array   | `tuple`             | p2m  |
+
+"p2m" indicates that conversion only applies from Python to Matlab 
+(from a Python `list` of numbers to a Matlab numeric array, and from a python `tuple` to a Matlab `cell` array).
+Other conversions apply in both directions (e.g. Matlab cell arrays are always converted to a Python `list` and vice versa).
+The conversion from a Matlab array to a Python array involves no data copies - the raw memory pointed to by Matlab is wrapped in a numpy `array` object.
+This is possible because that memory is only used in the Python function to be executed and not more widely.
+The reverse conversion (from a `numpy` array to a Matlab array) involves a data copy
+because Matlab `mex` does not support a shared data wrapping over memory which is not allocated by Matlab.
+
+Note that Matlab objects, enumerations and sparse data types are currently not converted and will give an "`Unrecognised input`" error.
+Likewise, Python objects or other data types not listed above returned by the Python functions will give an "`Unknown return type`" error.
+In principle both these objects can be wrapped using the mechanisms described above for Matlab objects (through a `MatlabProxyObject`)
```

Comment:
There appear to be 6 input types described here, so "both" is not appropriate. Did you mean "both these conversion directions"?

Reply:
Yes.